src/joystick/linux/SDL_sysjoystick.c

```diff
@@ -260,18 +260,20 @@ static int IsJoystick(const char *path, int fd, char **name_return, SDL_Joystick
     struct input_id inpid;
     char *name;
     char product_string[128];
+    SDL_UDEV_deviceclass class = 0;
 
-    if (ioctl(fd, JSIOCGNAME(sizeof(product_string)), product_string) >= 0) {
-        SDL_zero(inpid);
+    SDL_zero(inpid);
 #ifdef SDL_USE_LIBUDEV
-        SDL_UDEV_GetProductInfo(path, &inpid.vendor, &inpid.product, &inpid.version);
+    SDL_UDEV_GetProductInfo(path, &inpid.vendor, &inpid.product, &inpid.version, &class);
 #endif
-    } else {
+    if (ioctl(fd, JSIOCGNAME(sizeof(product_string)), product_string) <= 0) {
         /* When udev is enabled we only get joystick devices here, so there's no need to test them */
-        if (enumeration_method != ENUMERATION_LIBUDEV && !GuessIsJoystick(fd)) {
+        if ( enumeration_method != ENUMERATION_LIBUDEV &&
+             !(class & SDL_UDEV_DEVICE_JOYSTICK) && !GuessIsJoystick(fd) ) {
             return 0;
         }
 
+        /* The old code never used udev on the ev branch, so we still let ev overwrite any udev info */
```

Comment:
Is this comment going to make sense to a future maintainer? Talking about "the old code" seems like something that I would expect to see in a commit message justifying why the new behaviour is the way it is, rather than a comment in the code. Ideally the implementation should make sense on its own, without any history.

The non-legacy Linux 2+ kernel interface for input event devices is called evdev, and I think we should be consistent about referring to it that way, rather than sometimes saying evdev and sometimes ev. The reason we need to mention evdev at all, rather than having it be implicit - "well, this is Linux and it's an input device, so *obviously* it's evdev" - is that in addition to evdev, SDL has some limited support for the legacy Linux 0.x/1.x joystick interface via `/dev/input/js*`.

I'm also unsure what you mean by "let ev overwrite any udev info". All this function call is doing is: it's retrieving the product information (vendor, product, revision) directly from the device node, bypassing udev. If the two are contradictory, then there is a bug, because udev tags devices with its own properties (like `ID_JOYSTICK` which is the one you're interested in) but should never be unilaterally overwriting factual information from the kernel layer. The only reason why udev provides this duplicate information is that if it's available through udev, applications can retrieve it without needing to open the actual device node - but for SDL's use case, we want to open the device node anyway, so we might as well go directly to the canonical source.

In particular, the comment reads as though you're saying that this ioctl will overwrite our idea of the device class, which it specifically does not.

Reply:
I have updated the comment now to read
> Could have vendor and product already from udev, but should agree with evdev